content/tutorials/numpy_integration/grass_numpy_integration.qmd

```diff
@@ -0,0 +1,253 @@
+grid.add_field("topographic__elevation", fractal.ravel(), at="node")
+```
+
+Run the erosion of the landscape, extract the resulting NumPy array and visualize it:
+
+```{python}
+from landlab.components import LinearDiffuser
+from landlab.components import FlowAccumulator, FastscapeEroder
+
+fa = FlowAccumulator(grid, flow_director="D8")
+fsp = FastscapeEroder(grid, K_sp=0.01, m_sp=0.5, n_sp=1.0)
+ld = LinearDiffuser(grid, linear_diffusivity=1)
+
+for i in range(100):
+    fa.run_one_step()
+    fsp.run_one_step(dt=1.0)
+    ld.run_one_step(dt=1.0)
+
+elevation = grid.at_node['topographic__elevation'].reshape(grid.shape)
+
+plt.figure()
+plt.imshow(elevation, cmap='terrain', origin='upper')
+plt.colorbar(label='Elevation')
+plt.title('Topography after erosion')
+plt.xlabel('X-coordinate')
+plt.ylabel('Y-coordinate')
+plt.show()
+```
+
+![Eroded fractal surface with Landlab](fractal_landlab.webp)
+
+# From Landlab to GRASS
+
+Now we will bring the eroded topography back to GRASS for additional hydrology modeling.
+We will derive streams using the [r.watershed](https://grass.osgeo.org/grass-stable/manuals/r.watershed.html) and [r.stream.extract](https://grass.osgeo.org/grass-stable/manuals/r.stream.extract.html) tools.
+
+{{< fa wand-magic-sparkles >}} The `Tools` API allows us to directly plugin the NumPy `elevation` array into the tool call. {{< fa wand-magic-sparkles >}}
+
+```{python}
+tools.r_watershed(elevation=elevation, accumulation="accumulation")
+tools.r_stream_extract(
+    elevation=elevation,
+    accumulation="accumulation",
+    threshold=300,
+    stream_vector="streams",
+)
+
+```
+
+And visualize them using `gj.Map` on top of shaded relief:
+
+```{python}
+tools.r_relief(input=elevation, output="relief")
+
+m = gj.Map(width=400)
+m.d_rast(map="relief")
+m.d_vect(map="streams", type="line", color="blue", width=2)
+m.show()
+```
+
+![Streams derived from eroded topography in GRASS](streams.webp)
+
+Now if we want to store the eroded topography as a native GRASS raster, we can use
+[grass.script.array](https://grass.osgeo.org/grass-stable/manuals/libpython/grass.script.html#script.array.array)
+to create a GRASS array object with the dimensions given by the current computational region.
+Then we copy the NumPy array and write the raster map as GRASS native raster.
+
+```{python}
+# Create a new grasss.script.array object
+grass_elevation = ga.array()
+# Copy values from elevation array
+grass_elevation[...] = elevation
+# Write new GRASS raster map 'elevation'
+grass_elevation.write("elevation")
+```
+
+::: {.callout-note title="What about N-dimensional arrays?"}
+
+For 3D arrays, you can use [grass.script.array3d](https://grass.osgeo.org/grass-stable/manuals/libpython/grass.script.html#script.array.array3d) package.
+
+For N-dimensional arrays, check out the new [xarray-grass](https://github.com/lrntct/xarray-grass) bridge developed by [Laurent Courty](https://github.com/lrntct) {{< fa rocket >}}
+:::
+
```

Comment:
What about finishing with the 3D rendering which is in the thumbnail picture. That would actually make use of the converted array.